File: 프로그래머스/Level1/실패율.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool gg(pair<int, double> a, pair<int, double> b){
    if(a.second == b.second) return a.first < b.first;
    else return a.second > b.second;
}

vector<int> solution(int N, vector<int> stages) {
    vector<int> answer;
    vector<pair<int, double>> sil;

    for(int i=1; i<=N; i++){
        
        int nocl = 0;
        int dodal = 0;
        
        for(int j=0; j<stages.size(); j++){
            
            if(i == stages[j]) nocl ++;
            if(i <= stages[j]) dodal ++;  
            
        }
        
        if(dodal == 0 || nocl == 0) sil.push_back({i,0});
        else sil.push_back({i, (double)nocl/dodal});
        
    }
    
    sort(sil.begin(), sil.end(), gg);
    
    for(int i=0; i<sil.size(); i++){
        answer.push_back(sil[i].first);
    }
    
    return answer;
}

File: 프로그래머스/Level1/실패율.js
function solution(N, stages) {
    let tmpArr = new Array(N+1).fill(0);
    const len = stages.length;
    
    stages = stages.sort((a,b) => a-b);
    
    for(let i=0; i<len; i++) tmpArr[stages[i]-1] ++;

    let user = len;
    for(let i=0; i<N+1; i++){
        const unclearedUser = tmpArr[i];
        tmpArr[i] = [i+1, unclearedUser/user];
        user -= unclearedUser;
    }
    tmpArr = tmpArr.slice(0, tmpArr.length-1).sort((a, b)=> b[1]-a[1]);

    return tmpArr.map(e => e[0]);
}
